### PR TITLE
Show actual branch name in "Update From Default Branch" menu item

### LINKED ISSUE #3018: "Update From Default Branch" should use the default branch's name
_From https://github.com/desktop/desktop/issues/3016#issuecomment-335913972_

This is something I've run into too. For me, my eyes look for `master` and don't find it.

### PR DESCRIPTION
**Pull request by a-golovanov**
Fixes #3018

### CODE REVIEW

File: app/src/lib/stores/app-store.ts [hunk 1]
```diff
@@ -1424,38 +1424,48 @@ export class AppStore extends TypedBaseStore<IAppState> {
...
-      return defaultPRLabel
+  private getBranchesState(repository?: Repository) {
+    if (!repository || !repository.gitHubRepository) {
+      return undefined
     }
 
     const state = this.repositoryStateCache.get(repository)
 
     const { branchesState } = state
+    return branchesState
```
Comment (shiftkey) [inline app/src/lib/stores/app-store.ts]: :art: we can merge this into the previous line to create a one-liner

```ts
return state.branchesState
```

File: app/src/lib/stores/app-store.ts [hunk 2]
```diff
@@ -1424,38 +1424,48 @@ export class AppStore extends TypedBaseStore<IAppState> {
...
+  }
+
+  private getPullRequestLabel(repository?: Repository) {
+    const branchesState = this.getBranchesState(repository)
+    if (!branchesState) {
```
Comment (shiftkey) [inline app/src/lib/stores/app-store.ts]: We like to be explicit with these sorts of checks, rather than lean on JS truthiness too much

```suggestion
    if (branchesState == null) {
```

File: app/src/lib/stores/app-store.ts [hunk 3]
```diff
@@ -1424,38 +1424,48 @@ export class AppStore extends TypedBaseStore<IAppState> {
...
     return __DARWIN__ ? 'Show Pull Request' : 'Show &pull request'
   }
 
+  private getDefaultBranchName(repository?: Repository) {
+    const branchesState = this.getBranchesState(repository)
+    if (!branchesState) {
+      return undefined
+    }
+
+    const defaultBranch = branchesState.defaultBranch
```
Comment (shiftkey) [inline app/src/lib/stores/app-store.ts]: This one we can use object destructuring to simplify:

```suggestion
    const { defaultBranch } = branchesState
```

File: app/src/lib/stores/app-store.ts [hunk 4]
```diff
@@ -1424,38 +1424,48 @@ export class AppStore extends TypedBaseStore<IAppState> {
...
-      return defaultPRLabel
+      return undefined
     }
 
     return __DARWIN__ ? 'Show Pull Request' : 'Show &pull request'
   }
 
+  private getDefaultBranchName(repository?: Repository) {
+    const branchesState = this.getBranchesState(repository)
+    if (!branchesState) {
```
Comment (shiftkey) [inline app/src/lib/stores/app-store.ts]:
```suggestion
    if (branchesState == null) {
```

File: app/src/lib/stores/app-store.ts [hunk 5]
```diff
@@ -1424,38 +1424,48 @@ export class AppStore extends TypedBaseStore<IAppState> {
...
+      return undefined
+    }
+
+    const defaultBranch = branchesState.defaultBranch
+    return (defaultBranch && defaultBranch.upstreamWithoutRemote) || undefined
```
Comment (shiftkey) [inline app/src/lib/stores/app-store.ts]: Similarly, I'd like this to be a bit easier to read:

```ts
if (
  defaultBranch === null ||
  defaultBranch.upstreamWithoutRemote === null
) {
  return undefined
}

return defaultBranch.upstreamWithoutRemote
```

File: app/src/main-process/menu/build-default-menu.ts [hunk 6]
```diff
@@ -16,12 +16,21 @@ const defaultShellLabel = __DARWIN__
...
+  pullRequestLabel?: string
+  defaultBranchName?: string
+}
+
+export function buildDefaultMenu({
+  editorLabel = defaultEditorLabel,
+  shellLabel = defaultShellLabel,
+  pullRequestLabel = defaultPullRequestLabel,
+  defaultBranchName = defaultBranchNameValue,
+}: MenuLabels = {}): Electron.Menu {
```
Comment (shiftkey) [inline app/src/main-process/menu/build-default-menu.ts]: I think we can do away with the `= {}` usage in here and ensure `main.ts` line 203 passes in an empty object:

```ts
  let menu = buildDefaultMenu({})
```

File: app/src/lib/stores/app-store.ts (hunk 2 above)
Comment (a-golovanov) [reply app/src/lib/stores/app-store.ts]: Is there any policy/recomendation regarding == and === operators?

Comment (a-golovanov) [reply app/src/lib/stores/app-store.ts]: Interesting, I would love to know the reasons besides these rules. In any case, I will try to follow them.

File: app/src/main-process/menu/build-default-menu.ts (hunk 6 above)
Comment (a-golovanov) [reply app/src/main-process/menu/build-default-menu.ts]: I there any reasoning behind this recommendation besides improved readability? Readability is a subjective thing and sort of hard to follow.

File: app/src/lib/stores/app-store.ts (hunk 2 above)
Comment (a-golovanov) [reply app/src/lib/stores/app-store.ts]: Sorry, I messed up commit description, I'm new to all of this :)

**Comment by tierninho** [comment]
The UI was tested for this change, and overall it functions well. 

Noticed there is not limit to the truncation in the menu item. This should probably be capped at 25 characters (followed by `...`) to match the truncation we use in the branch list.

screen shot 2018-10-19 at 11 51 26 am

- [x] actual name of default branch is shown in menu item
- [x] changing default name updates the menu item
- [x] user cannot rename default branch
- [x] default branch remains greyed out if user is already on default branch
- [x] keyboard cmd still intact

Build/OS: 1.4.4-beta0, mac, dev 6d847e853143ed40a0bcef7e9b617a0a1c04288e

**Comment by shiftkey** [comment]
@tierninho that's a great find! @a-golovanov are you able to incorporate that?

**Comment by shiftkey** [comment]
> 6d847e853143ed40a0bcef7e9b617a0a1c04288e

@tierninho Just a heads up, this commit ID isn't being linked to in your comment because it doesn't exist on GitHub (so I can't get to it from my machine and "see what you're seeing". For example, I can checkout ddc8145db655dec67b02f55349bf4adfcd6d4296 which is the current tip of the PR and by mentioning it here I see it light up.

File: app/src/main-process/menu/build-default-menu.ts (hunk 6 above)
Comment (shiftkey) [reply app/src/main-process/menu/build-default-menu.ts]: We generally try to avoid optional parameters with defaults in TS, because it can be used to obscure what the API is capable of. There's already a fair bit of complexity here around destructuring this parameter and providing default values for it, so I settled on requiring the caller to provide an empty object.

If there were lots of callers that needed to do this I'd likely spend more time figuring out the right API to support them, but as it's one caller (and called exactly once during startup) I was fine with dropping the optional part.

**Comment by a-golovanov** [comment]
> @tierninho that's a great find! @a-golovanov are you able to incorporate that?

I can implement truncation to the fixed number of characters, but I cannot implement it the same way it's done in the branch list. The branch list truncates names with CSS, so the number of visible characters depends on characters widths and width of the list itself.
Should I proceed with truncation to the fixed number of characters even if it will look different comparing to the branch list?

**Comment by shiftkey** [comment]
> Should I proceed with truncation to the fixed number of characters even if it will look different comparing to the branch list?

@a-golovanov this is fine. The branch name should not expand the menu too much, and _how_ the branch list is currently trimmed is a good example. We obviously can't use the same implementation here, but we should try and be consistent.

**Comment by tierninho** [comment]
Truncation UI change = ✅

File: app/src/lib/truncate-with-ellipsis.ts [hunk 7]
```diff
@@ -0,0 +1,32 @@
+/** Truncate a single line unicode string by a given maxLength and add ellipsis if necessary */
+export function truncateWithEllipsis(str: string, maxLength: number) {
```
Comment (shiftkey) [inline app/src/lib/truncate-with-ellipsis.ts]: This seems a great chance to write some tests for this, especially now that we have to be aware of Unicode characters in the string. Are you able to add some tests to your PR using the guide I started in https://github.com/desktop/desktop/pull/5983? I'd love some feedback on if that helps, or what else we can add.

File: app/src/main-process/menu/build-default-menu.ts (hunk 6 above)
Comment (shiftkey) [reply app/src/main-process/menu/build-default-menu.ts]: @a-golovanov also, here's some extra context about why we like to avoid these implicit parameters: https://github.com/desktop/desktop/pull/5979#discussion_r227826267

File: app/src/lib/truncate-with-ellipsis.ts (hunk 7 above)
Comment (a-golovanov) [reply app/src/lib/truncate-with-ellipsis.ts]: I wrote tests, see `app/test/unit/truncate-with-ellipsis-test.ts`. I used `chai` though following other tests as example. I'll migrate my tests to `jest` today.
